rustls/examples/internal/test_ca.rs

```diff
@@ -0,0 +1,365 @@
+        &PKCS_RSA_SHA512,
+        &PKCS_ECDSA_P256_SHA256,
+        &PKCS_ECDSA_P384_SHA384,
+        &PKCS_ECDSA_P521_SHA512,
+        &PKCS_ED25519,
+    ];
+
+    let mut certified_keys = HashMap::with_capacity(roles.len() * sig_algs.len());
+    for role in roles {
+        for alg in sig_algs {
+            // Generate a key pair and serialize it to a PEM encoded file.
+            let key_pair = keypair_for_alg(alg);
+            let mut key_pair_file = File::create(role.key_file_path(alg)).unwrap();
+            key_pair_file
+                .write_all(key_pair.serialize_pem().as_bytes())
+                .unwrap();
+
+            // Issue a certificate for the key pair. For trust anchors, this will be self-signed.
+            // Otherwise we dig out the issuer and issuer_key for the issuer, which should have
+            // been produced in earlier iterations based on the careful ordering of roles.
+            let cert = match role {
+                Role::TrustAnchor => role
+                    .params(alg)
+                    .self_signed(&key_pair)
+                    .unwrap(),
+                Role::Intermediate => {
+                    let CertifiedKey {
+                        cert: issuer,
+                        key_pair: issuer_key,
+                    } = certified_keys
+                        .get(&(Role::TrustAnchor, alg))
+                        .unwrap();
+                    role.params(alg)
+                        .signed_by(&key_pair, issuer, issuer_key)
+                        .unwrap()
+                }
+                Role::EndEntity | Role::Client => {
+                    let CertifiedKey {
+                        cert: issuer,
+                        key_pair: issuer_key,
+                    } = certified_keys
+                        .get(&(Role::Intermediate, alg))
+                        .unwrap();
+                    role.params(alg)
+                        .signed_by(&key_pair, issuer, issuer_key)
+                        .unwrap()
+                }
+            };
+            // Serialize the issued certificate to a PEM encoded file.
+            let mut cert_file = File::create(role.cert_pem_file_path(alg)).unwrap();
+            cert_file
+                .write_all(cert.pem().as_bytes())
+                .unwrap();
+            // And to a DER encoded file.
+            let mut cert_file = File::create(role.cert_der_file_path(alg)).unwrap();
+            cert_file.write_all(cert.der()).unwrap();
+
+            certified_keys.insert((role, alg), CertifiedKey { cert, key_pair });
+        }
+    }
+
+    // Write the PEM chain and full chain files for the end entity and client certs.
+    // Chain files include the intermediate and root certs, while full chain files include
+    // the end entity or client cert as well.
+    for role in [Role::EndEntity, Role::Client] {
+        for alg in sig_algs {
+            let CertifiedKey { cert: root, .. } = certified_keys
```

Comment:
Is all the destructuring really worth it? I find these vertically squashed together three-line destructuring expressions kind of hard to read, and not sure they buy us a lot especially in the case where we're only taking a single field.

Compare:

```rust
            let CertifiedKey { cert: root, .. } = certified_keys
                .get(&(Role::TrustAnchor, alg))
                .unwrap();
            let CertifiedKey {
                cert: intermediate, ..
            } = certified_keys
                .get(&(Role::Intermediate, alg))
                .unwrap();
```

To

```rust
let root = certified_keys.get(&(Role::TrustAnchor, alg)).unwrap().cert;
let intermediate = certified_keys.get(&(Role::Intermediate, alg)).unwrap().cert;
```

(Don't know if rustfmt will actually allow these to be on one line...)

Reply:
I removed the destructuring. You're right, it was noisey and wasn't providing much value.